### PR TITLE
AP_Scripting: mount-djirs2 script handles latest gimbal software

### PR DESCRIPTION
**Pull request by rmackay9**
This is a replacement for PR https://github.com/ArduPilot/ardupilot/pull/24461

Three important fixes are included:

1. Support for both legacy and latest gimbal firmware
2. bug fix the reported yaw angle when the gimbal is mounted "upside down"
3. reduce the reply timeout from 1 sec to 0.1 sec.  Replies are normally received within 5ms and waiting for 1sec leads to the gimbal stopping

This has been tested on a DJIR2 running the legacy firmware.  I hope that @tpwrules can check that it also works on the latest firmware.

### CODE REVIEW

**Comment by tpwrules** [comment]
I'll look into this later this week. Thanks for getting back to it.

> the gimbal is mounted "upside down" which means the camera is above the gimbal.

I don't think this is true? In all the marketing material and documentation and menus, the "normal" orientation of the gimbal has the camera above the handle and control unit. In most drone applications the camera will hang below the handle and control unit, which is why we provide the upside down mode, correct?

As for the roll/pitch being reversed, I'm not sure what's up. The code to retrieve the data as it's written directly contradicts the documentation at the top of the file and the spec [available here](https://terra-1-g.djicdn.com/851d20f7b9f64838a34cd02351370894/Ronin%E7%B3%BB%E5%88%97/DJI_R_SDK_Protocol_and_User_Interface_EN_v2.3.pdf). But if it works for you there must be some gimbals which are reversed? I'll double check this too.

Might be best to just make everyone update.

**Comment by rmackay9** [comment]
@tpwrules,

Txs for the feedback and offer of testing.  You may be right about the definition of "upside-down". I'll check with another user I know about what he thinks is rightside-up vs upside-down.

**Comment by tpwrules** [comment]
Regarding the testing script in the branch in the old issue, I get `poscontrol len: 19`, `att len: 26`. These sound like they match expectations.

The script in this branch works fine with my gimbal. But the roll and pitch axes are for sure swapped. If I set `DJIR_DEBUG=2` and move my controller's roll input (function 212), then the gimbal roll axis moves, but the pitch number changes in the debug output. Same thing if I move function 213, then the pitch axis moves, but the roll number is what changes. Yaw seems okay.

Again, the protocol documentation contained within the script itself and the [PDF from DJI](https://terra-1-g.djicdn.com/851d20f7b9f64838a34cd02351370894/Ronin%E7%B3%BB%E5%88%97/DJI_R_SDK_Protocol_and_User_Interface_EN_v2.3.pdf) suggests the interpretation my gimbal uses is correct and the script code is wrong. You say your gimbal moves correctly, did you check that the reported data is correct using the debug option? Can you remind which firmware your gimbal is on? Mine is 01.05.00.20.

**Comment by rmackay9** [comment]
@tpwrules,

Thanks for the testing and feedback.  As hard as it is to believe, I think DJI had a bug in the angle reporting for roll and pitch in the legacy firmware (they were swapped).  I've updated the PR so that for the "latest" DJI firmware it uses the correct order.

Maybe you could comment whether you're happy with these changes?  That'll clear the way for merging I think.  Txs again.

File: libraries/AP_Scripting/drivers/mount-djirs2-driver.lua [hunk 1]
```diff
@@ -653,12 +656,27 @@ function parse_byte(b)
...
 
           -- parse attitude reply message
-          if (expected_reply == REPLY_TYPE.ATTITUDE) and (parse_length >= 20) then
-            local ret_code = parse_buff[13]
+          if (expected_reply == REPLY_TYPE.ATTITUDE) and (parse_length >= ATTITUDE_PACKET_LEN.LEGACY) then
+            -- default to legacy format but also handle latest format
+            local ret_code_field = 13
+            local yaw_field = 15
+            local pitch_field = 17
+            local roll_field = 19
```
Comment (tpwrules) [inline libraries/AP_Scripting/drivers/mount-djirs2-driver.lua:664]: Would like to see a comment here that this field ordering contradicts the spec but is how an actual gimbal of version XX (whichever version yours is) is observed to behave.

**Comment by tpwrules** [comment]
I personally would like to see support for the legacy protocol dropped completely. Updating is slow but not hard; the user has to have hooked up the app to activate the gimbal in the first place. We are also assuming whichever version updated the packet length is whichever version fixed the reporting bug, which is hard to determine.

Aside from that comment nit, I'm happy with these changes if you're happy with the additional complications of supporting multiple versions.

**Review by peterbarker — APPROVED** [review]

**Comment by rmackay9** [comment]
@tpwrules,

I think it's so easy for us to keep supporting the old firmware that we should.

**Comment by peterbarker** [comment]
@tpwrules Randy has committed to adding coimments on the roll/pitch issue in a follow-up PR.

We discussed removing legacy protocol support but Randy made a spirited argument that in this case it's a very small amount of work to support the older firmware and there's no other drawback in doing so.

We do enforce minimum firmware version on some camera drivers, but in this case we don't need to - yet!

Thanks for looking through this PR, we do appreciate the reviews :-)

**Pull request merged**